README
                             Halo_Radio 


             Copyright (c) 2001-2005, Kremlor Industries

------------------------------------------------------------------------
    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc.:
        59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
	
    For more details see the file COPYING.

------------------------------------------------------------------------

[This software is maintained by phil@kremlor.net]

What is halo_radio ?
--------------------

halo_radio is an automatic dj system for your mp3 collection. An icecast
source daemon, a fully featured web interface, a back end MySQL database,
and back end utilities to manage a web-based mp3 radio station. 

the web interface allows for searching, making requests, managing the
request queue, viewing recent activity, historical statistics for usage,
and more. statistical analasys of requests and kills on a per-song basis,
allow the system to automaticly maintain a a base playlist of popular songs.

A backed library to the database makes python based clients trivial to
code. The biggest problem now is documentation.


Troubleshooting / Questions / Rants :
--------------------------------------

    http://kremlor.net/projects/halo_radio/

    http://kremlor.net/mailman/listinfo/halo_radio


Software Prerequisites
----------------------

HaloRadio makes use of existing software to provide some parts of its
features and functionality.  For example the transport of mp3s is 
provided by Icecast and the backed database is provided by MySQL.  
There are quite a few prerequisite applications to be installed before 
your streaming radio station is fully operational.  While the author 
would like to provide installation and configuration help for all the
software used in this project these issues are extensively covered by
their respective maintainers.

The author suggests recent versions of all programs listed below or
a nice package management system like debian's apt.

    Required -

    Python: http://www.python.org/
    SimpleTAL: http://www.owlfish.com/software/simpleTAL/
    Mysql: http://www.mysql.org/
    MySQL-python: http://sourceforge.net/projects/mysql-python
    IceCast: http://www.icecast.org/
    libshout-2.x: http://www.icecast.org/download.php
    shout-python: http://www.icecast.org/download.php
    Apache: http://www.apache.org/
    Lame: http://lame.sourceforge.net/

    Optional -

    RRDTool: needed for listener stat graphs.

All of the mentioned software is available in debian packages except lame
and djb's daemontools. those two packages are available on my debian
repository ( see http://kremlor.net/~phil for more info ).

Manual Installation
------ ------------
To install as a user without root access see: Manual Installation (non-root)

1) Create a local user:

    # groupadd radio
    # useradd -d /usr/share/haloradio -g radio radio

2) UN-tar your source distribution:

    # tar -zxvf haloradio-xxx.tar.gz

3) Install the package:

    # cd haloradio-xxx
    # make install

3)  Setup your radio database:

    Create a radio database:
    mysql> create database radio;

    Create a database user:
    mysql> grant all on radio.* to radio@localhost  \
    identified by 'password';

    Populate your radio database:
    # mysql -u radio -p radio <  \
    /usr/share/haloradio/halo_radio.mysql

4) Edit your Configuration file (HaloRadio.ini)

    $ vi /etc/haloradio/HaloRadio.ini

    ( if you're upgrading the new config will be installed as
     /etc/haloradio/HaloRadio.ini-dist)

6) Sync the file system with the :

    Run DBSync.py to make halo_radio aware of your mp3s:
    # su - radio
    $ ./DBSync.py -cfsp

	( Later set up a cron job to update the system regularly : see
	 haloradio-xxx/debian/haloradio-backend.cron.d)


7) Create a crontab entry to sync your database (see
    haloradio-xxx/debian/haloradio-backend.cron.d)


8) Test the daemon:

    # su - radio
    $ ./radiod.py

9) Setup the daemon in Daniel J. Bernstein's daemontools:
        
    Install daemontools ( see http://cr.yp.to/daemontools/install.html)

    Run the service directory install script
    # cd /usr/share/haloradio
    # sh ./radio-conf.sh

    Link the supervise directory to the service directory
    # ln -s /etc/radio /service/radio

    The daemon should start. Check /etc/haloradio/svsdir/log/main/current
     for logs.

9) Configure apache for web interface:

    Configure your web servers document root to point to your
    Halo_Radio installation directory. An apache virtual entry
    might look something like this:

    <VirtualHost radio.mydomain.com:80>
        
        DocumentRoot /usr/share/haloradio/public_html
        ServerName radio.mydomain.com
    </VirtualHost>

    Notes
    -----
        - Your DocumentRoot should be separate from your
	  install directory (Hi Ryan!)
	- the install will copy a .htaccess into your
	  DocumentRoot which may need to be edited

To use the web interface:

    Point your browser at http://radio.mydomain.com/

    help lives here : http://radio.mydomain.com/help

    At this point you should have a fully functional installation of
    HaloRadio.   

To create your admin user:

    use the MySQL monitor to manually add a row to the database for
    for your admin user. see the bottom of halo_radio.mysql for an example.

To create your m3u file:

    To make it easier for others to listen radio broadcast you can 
    create a m3u file that contains the following line:

    http://radio.mydomain.com:8000/radio

To connect with your mp3 client: 

    Use xmms, winamp, mplayer etc. 

    $ mplayer http://radio.mydomain.com:8000/radio
   
If the installation of Halo_Radio failed, check whether all the required
software is correctly installed. If it still failed, please send bug
report to the halo_radio mailing list mentioned near the beginning of
this document.


Manual Installation (non-root)
------ ------------ ----------

1) UN-tar your source distribution:

    # tar -zxvf haloradio-xxx.tar.gz

2) Install the package:

    # cd haloradio-xxx
    # make DESTDIR=~/haloradio install-user

3)  Setup your radio database:

    Create a radio database:
    mysql> create database radio;

    Create a database user:
    mysql> grant all on radio.* to radio@localhost  \
    identified by 'password';

    Populate your radio database:
    # mysql -u radio -p radio <  \
    ~/haloradio/usr/share/haloradio/halo_radio.mysql

4) Edit your Configuration file (HaloRadio.ini)

    $ vi ~/haloradio/etc/haloradio/HaloRadio.ini

    ( if you're upgrading the new config will be installed as
     ~/haloradio/etc/haloradio/HaloRadio.ini-dist)

5) Sync the file system with the :

    Run DBSync.py to make halo_radio aware of your mp3s:
    $ cd ~/haloradio/usr/share/haloradio
    $ ./DBSync.py -cfsp

	( Later set up a cron job to update the system regularly : see
	 haloradio-xxx/debian/haloradio-backend.cron.d)


6) Create a crontab entry to sync your database (see
    haloradio-xxx/debian/haloradio-backend.cron.d)


7) Test the daemon:

    $ cd ~/haloradio/usr/share/haloradio
    $ ./radiod.py

8) Verify your apache config for web interface:

Refer to step 9 of the Manual Installation.


So, You're up and Running, What Now ?
-------------------------------------

- Turn halo_radio AutoAI mode on.

When you ran DBSync.py -p, you created some playlists. halo_radio uses
playlists to decide what to play when there are no requests in queue. You
can change your default playlist with an admin(a) user on the Admin page.
there are 4 playlists that DBSync.py builds:

	- Master -
		This contains all the valid mp3s in your archive.
	- Requested - 
		This playlist contains all songs that have been requested.
	- AutoAI -
		This playlist contains all the songs with a net positive
		popularity. (Popularity is the number of requests - the number
		of kills.

You've got DBSync running in cron, right ? well, then after you've requested
a bunch of songs, You can switch your playlist over to AutoAI. brand new
PlayLogic in this release (0.2.0pre3) also uses active users to select some
of it's random songs. see the ChangeLog for more info.





ChangeLog
- 0.2.5-4
  - database changes
	ALTER TABLE `users` ADD COLUMN `style` int(10) DEFAULT 1;
  	CREATE TABLE `styles` ( `id` int(10) unsigned NOT NULL auto_increment,
	 `style` tinytext NOT NULL, PRIMARY KEY (`id`) ) TYPE=MyISAM;
	INSERT INTO `styles` VALUES (1,'style.css');
	INSERT INTO `styles` VALUES (2,'no_style.css');
	INSERT INTO `styles` VALUES (3,'style_green.css');
	INSERT INTO `styles` VALUES (4,'style_pink.css');

- 0.2.0pre7
  - database changes
  	> CREATE TABLE relays ( id int(10) unsigned NOT NULL auto_increment,
	url text NOT NULL,title tinytext, comment text, added_date datetime
	default NULL, plays smallint(5) unsigned default '0', requests
	smallint(5) unsigned default '0', kills smallint(5) unsigned default
	'0', PRIMARY KEY  (id));
  	> INSERT INTO `config` VALUES ('','dbversion','0');
	> INSERT INTO `config` VALUES ('','current_relay','0');

  - source daemon
	- auto-normalization support disabled until bugs can be worked out.
  - logic changes
	- GetListeners changed to pull data from status2.xsl (included
	   with recent versions of icecast.
- 0.2.0pre6
  - database changes
  	> alter table request_queue add column hold int(1) unsigned NOT NULL
	-> Default 0;
  - config file changes ( see HaloRadio.ini-dist for details )
	- rerequest_limit=0
	- replay_limit=0
  - web interface changes
  	- added support to search multiple users requests. Advanced Search.
	- added holding / releasing of requests.
	- added support to view the held requests in the queue.
- 0.2.0pre5
  - database changes
  	> alter table songs add column scale FLOAT default 0;
  - web interface changes
  	- fixed a bug in the newUser signup that was sending messages from
	halo_radio@server.DOMAIN , now emails should come from
	halo_radio@hostname.
	- search function searches some id3 tags as well as the path and
	filename. (artist, album, title)

  - source daemon changes
  	- added auto-normalization support.

- 0.2.0pre4
  - database changes ( see halo_radio.mysql for details )
        - new table 'wall'. create it.
  - config file changes ( see HaloRadio.ini-dist for details )
	+ request_quota=2
	+ do_fortune=1
	+ active_user_weight=3
  - web interface changes
	- users over the Request Quota will silently fail adding requests
	- kill moved from main menu to current info. this should cut
	down on mistaken kills.
	- CurrentInfo page refreshes automaticly at next song.
	- Listeners page renamed to Activity. the page refreshes every ten
	minutes.
	- all verified users have access to browse other users stats.
	- userstats page allows user to enter selected lengths of lists.
  - source daemon changes
	- will now resample files that are not 44,100 samples per second.
	- active user weight for random songs tunable.
  - utility changes
	- DBSync.py now adds mp3's with samplerates other than 44.1 khz
	- halostatcron.py - updated graphs to more acturatly display
	historical data. also changed the colors of the graphs to make
	them more 'halo'.
	- filelist.py now helps you to create audio and mp3 cds based on
	specific users tastes. (more info to come)
	- rename.py added. this utility allows you to run search and
	replaces on songs' paths in the db (for when you move mp3s around
	in the archive and don't wanna loose stats)
- 0.2.0pre3
  - database changes
	> DELETE FROM config where name = "announce_delay";
  - config file changes, HaloRadio.ini
  	  [general]
	+ announce_path=halo_radio
	+ announce_delay=120
  - web interface changes
  	- added Admin page, with a playlist selector (for which playlist to
	select random songs from when there are no requests)
	- fixed newUser setup to not cc: me, instead use the config file admin
	email setting. :-)
	- updated/fixed/added a bunch of help and stuff.
	- songInfo page now shows per-user stats.
	- userInfo/MyInfo page shows top100 and bottom 25 songs for the user.
  - source daemon changes
  	- added user specific random song request for active users. 33% of the
	randomly played songs will use the active users to generate a
	temporary list of popular songs for random selection from.
- 0.2.0pre2
  - database changes:
  	mysql> alter table songs add column added_date DATETIME;
	mysql> create index added_date ON songs (added_date);
  - config file changes:
    - Changed:
    	general.stream_url: UNUSED REMOVE
	general.guest_has_kill: [0/1] guest can kill random or own requests
	general.guest_has_request: [0/1] guest can search and request
	general.guest_has_songnfo: [0/1] guest can list songnfo (d/l link)
	general.guest_has_signup: [0/1] guest can signup

    	shout.url: UNUSED REMOVE
    	shout.bitrate: now accepts comma separated list of bitrates


- halo_radio-snap-20050106205155 (0.2.0pre1)

  - geting ready for initial release. call this a pre-release. the fix
    up docs and lingering bugs before initial public release.


DBSync.py
#!/usr/bin/python
#
# DBSync.py - fs sync util
#    Copyright (C) 2004 Philip J Freeman
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#

from string import find

import os,sys,time,getopt

import HaloRadio
import HaloRadio.Song as Song
import HaloRadio.SongListMaker as SongListMaker
import HaloRadio.Playlist as Playlist
import HaloRadio.Config as Config
import HaloRadio.MP3Info as MP3Info


arc_root=HaloRadio.conf['path.root']
arc_len=len(arc_root)+1

def check_string( string ):
      if string == None:
              return 1
      if len(string) > 255:
              return 1
      return 0

def make_printable_string( String ):
	newString = ""
	import string
	for i in range(0,len(String)):
		if String[i] in string.printable:
			newString += String[i]
	return newString
	
def update_dbversion():
	cfg = Config.Config()
	current_version=1

	dbversion=int(cfg.GetConfigItem("dbversion"))

	if dbversion<1:
		c = HaloRadio.db.cursor()
		c._do_query( """
ALTER TABLE `users` ADD COLUMN `style` int(10) DEFAULT 1;
			""")
		c._do_query( """
CREATE TABLE `styles` (
  `id` int(10) unsigned NOT NULL auto_increment,
  `style` tinytext NOT NULL,
  PRIMARY KEY (`id`)
) TYPE=MyISAM;
			""")
		c._do_query( """

INSERT INTO `styles` VALUES (1,'style.css');
			""")
		c._do_query( """
INSERT INTO `styles` VALUES (2,'no_style.css');
			""")
		c._do_query( """
INSERT INTO `styles` VALUES (3,'style_green.css');
			""")
		c._do_query( """
INSERT INTO `styles` VALUES (4,'style_pink.css');
			""")

		cfg.SetConfigItem("dbversion","1")
	return None

def check_dbversion():
	cfg = Config.Config()
	current_version=1

	try:
		dbversion=cfg.GetConfigItem("dbversion")
	except:
		print "WARNING: dbversion tag not found."
		print "       : Check that your DB structure is up to date."
		return 0

	dbversion=int(dbversion)

	if dbversion < current_version:
		print "WARNING: dbversion tag indicates your database is out of date. Updating..."
		update_dbversion()

	return 1

def update_db(slow, verbose=0):  

	cfg = Config.Config()


	timenow = time.time()
	#print "timenow %s" % timenow

	if slow:
		filestr = "find %s -follow -type f -name \"*.mp3\"" % ( arc_root )
	else:
		try:
			lastupdatesecs=float(cfg.GetConfigItem("lastupdate"))
		except:
			lastupdatesecs=0

		#print "lastupdatesecs %s" % lastupdatesecs
		timediff = timenow - lastupdatesecs
		#print "timediff %s" % timediff

		filestr = "find %s -type f -cmin -%d -name \"*.mp3\"" % ( arc_root, (timediff/60)+1 )
	cfg.SetConfigItem("lastupdate",str(timenow))
	#print filestr
	files=os.popen(filestr)
	
	line_in = files.readline()
	
	while ( line_in != "" ):
		file = line_in[:-1]		# chomp the \n
		line_in = files.readline()	# grab the next line
		path = file[arc_len:]		# cut the root off the filename

		# Check to see if this file is in the 
		#log_message( verbose, "file_in: %s" % ( path ) )
		songlist = SongListMaker.SongListMaker()
		songlist.SearchPath( path )
		if ( len(songlist.list) > 1 ):
			for sid in songlist.list:
				song = Song.Song( sid )
				print "#%d:%s" % ( sid, song.path )
			raise "you got a problem here."
		try:
			nfo = MP3Info.MP3Info(open(file, 'rb'))
		except:
			print "#Error Parsing file: "
			print """ls -la "%s" """%( file )
			continue

		if nfo.mpeg.valid != 1:
			print "#Invalid MPEG in File: %s" % ( file )
			print """ls -la "%s" """%( file )
			continue
		artist = ""
		year = ""
		comment = ""
		genre = ""
		track = 0
		title = ""
		album = ""
		if nfo.artist != None:
			artist = make_printable_string(nfo.artist)
		if nfo.album != None:
			album = make_printable_string(nfo.album)
		if nfo.title != None:
			title = make_printable_string(nfo.title)
		if nfo.track != None:
			track = nfo.track
		if nfo.genre != None:
			genre = nfo.genre
		if nfo.comment != None:
			comment = make_printable_string(nfo.comment)
		if nfo.year != None:
			year = nfo.year
		mpeg_version = nfo.mpeg.version
		mpeg_bitrate = nfo.mpeg.bitrate
		mpeg_samplerate = nfo.mpeg.samplerate
		mpeg_length = int ((nfo.mpeg.length_minutes*60 ) + nfo.mpeg.length_seconds)
		mpeg_emphasis = nfo.mpeg.emphasis
		mpeg_mode = nfo.mpeg.mode
		#if mpeg_samplerate != 44100:
		#	print "#Invalid mpeg_samplerate %s in File: %s" % ( mpeg_samplerate, file )
		#	print """ls -la "%s" """%( file )
		#	continue
		if comment:
			if find(comment,"iTunes") != -1:
				comment = "iTunes comment not added to database"
			if len(comment) > 1024:
				comment = "over 1024 bytes truncating."
		try:
			track = int(track)
		except:
			track = 0

		if ( len(songlist.list) > 0 ):
			song = Song.Song( songlist.list[0] )
			if (song.artist == artist) and (song.album== album) and (song.title== title) and (int(song.track) == int(track)) and (song.genre == genre) and (song.comment == comment) and (song.year == year) and ( song.mpeg_version == mpeg_version ) and ( int(song.mpeg_bitrate) == int(mpeg_bitrate) ) and ( int(song.mpeg_samplerate) == int(song.mpeg_samplerate) )and (int(song.mpeg_length) == int(mpeg_length) ) and (song.mpeg_emphasis == mpeg_emphasis) and (song.mpeg_mode == mpeg_mode):
				continue
			else:
				log_message( verbose, 
					"fl:%s:%s:%s:%s:%s:%s:%s:%s:%s:%s:%s:%s:%s" % (
					artist,album,title,track,genre,comment,year,mpeg_version,mpeg_bitrate,mpeg_samplerate,mpeg_length,mpeg_emphasis,mpeg_mode )
				)
				log_message( verbose,
					"db:%s:%s:%s:%s:%s:%s:%s:%s:%s:%s:%s:%s:%s" % (
						song.artist,song.album,song.title,song.track,song.genre,song.comment,song.year,song.mpeg_version,song.mpeg_bitrate,song.mpeg_samplerate,song.mpeg_length,song.mpeg_emphasis,song.mpeg_mode )
					)
				print "#Updating", path
				song.Update(artist, album, title, track, genre, comment, year, mpeg_version, mpeg_bitrate, mpeg_samplerate, mpeg_length, mpeg_emphasis, mpeg_mode )
				continue
		
		print "#Adding", path
		song = Song.Song( 0, path, artist, album, title, track, genre, comment, year, mpeg_version, mpeg_bitrate, mpeg_samplerate, mpeg_length, mpeg_emphasis, mpeg_mode )
		log_message( verbose,
			"ad:%s:%s:%s:%s:%s:%s:%s:%s:%s:%s:%s:%s:%s" % (
				song.artist,song.album,song.title,song.track,song.genre,song.comment,song.year,song.mpeg_version,song.mpeg_bitrate,song.mpeg_samplerate,song.mpeg_length,song.mpeg_emphasis,song.mpeg_mode )
			)

def clean_db(verbose=0):

		slm = SongListMaker.SongListMaker()
		slm.GetAll()
		for songid in slm.list:
			song = Song.Song(songid)
			if ( os.access(arc_root + "/" + song.path, os.R_OK)==0 ):
				song.Delete()
				print "#Deleted", song.path, "from the database"

def log_message(ifval,msg):
	if ifval:
		print "#",msg

def make_master_playlist():
	try:
		playlist = Playlist.Playlist( 1 )
	except:
		playlist = Playlist.Playlist( 0, "Master" )

	# delete any songs from playlist 1
	playlist.Clear()
	# add all songs to playlist 1
	all_songs = SongListMaker.SongListMaker( )
	all_songs.GetAll()
	for songid in all_songs.list:
		playlist.AddSong( songid )

def make_request_playlist():
	try:
		playlist = Playlist.Playlist( 2 )
	except:
		playlist = Playlist.Playlist( 0, "Requested" )

	playlist.Clear()
	reqested_songs = SongListMaker.SongListMaker( )
	reqested_songs.GetWhere("requests > 0")
	for songid in reqested_songs.list:
		playlist.AddSong( songid )


def make_smart_auto_playlist():
	try:
		playlist = Playlist.Playlist( 3 )
	except:
		playlist = Playlist.Playlist( 0, "AutoAI" )

	playlist.Clear()
	reqested_songs = SongListMaker.SongListMaker( )
	reqested_songs.GetWhere("requests > kills")
	for songid in reqested_songs.list:
		playlist.AddSong( songid )

def make_announce_playlist():
	try:
		playlist = Playlist.Playlist( 4 )
	except:
		playlist = Playlist.Playlist( 0, "Announce" )
	playlist.Clear()
	slm = SongListMaker.SongListMaker( )
	slm.GetWhere("""path LIKE "%s%%" """%(HaloRadio.conf['general.announce_path']))
	for songid in slm.list:
		playlist.AddSong( songid )

def usage():
	print
	print "%s: halo radio database sync'r" % ( os.path.basename(sys.argv[0] ) )
	print "	-h, --help	this cruft"
	print " -v, --verbose   print what we're doing"
	print "	-c, --clean	clean db"
	print "	-f, --scanfs	scan archive"
	print "	-s, --slow	slow search of fs"
	print " -p, --playlists	build playlists"
	print


clean=0
slow=0
verbose=0
scanfs=0
playlists=0

check_dbversion()

try:
	opts, args = getopt.getopt( sys.argv[1:], "hcvsfp", [ "help", "clean", "verbose", "slow", "scanfs", "playlists" ])
except getopt.GetoptError:
	usage()
	sys.exit(2)

for o, a in opts:
	if o in ("-h", "--help"):
		usage()
		sys.exit()
	if o in ("-v", "--verbose"):
		verbose = 1
	if o in ("-c", "--clean"):
		clean = 1
	if o in ("-s", "--slow"):
		slow = 1
	if o in ("-f", "--scanfs"):
		scanfs = 1
	if o in ("-p", "--playlists"):
		playlists = 1
	#if o in ("-", "--"):
	#	 = 1


if not clean and not scanfs and not playlists:
	usage()

if clean:
	log_message( 1, "checking files in db against disk...")
	clean_db(verbose)

if scanfs:
	log_message( 1, "checking files on disk against db...")
	update_db(slow, verbose)

if playlists:
	log_message( 1, "building playlists...")
	make_master_playlist()
	make_request_playlist()
	make_smart_auto_playlist()
	make_announce_playlist()

